src/nft.sol

```diff
@@ -13,185 +13,156 @@
+        assetManager = AssetManagerLike(assetManager_);
+        identityFactory = IdentityFactoryLike(identityFactory_);
+        keyManager = KeyManagerLike(keyManager_);
     }
 
-    event Minted(address usr, uint tkn);
+    event Minted(address to, uint tokenId);
 
-    // --- Utils ---
-    function concat(bytes32 b1, bytes32 b2) pure internal returns (bytes memory) {
-        bytes memory result = new bytes(64);
-        assembly {
-            mstore(add(result, 32), b1)
-            mstore(add(result, 64), b2)
-        }
-        return result;
+    /**
+     * @dev Mints a token to a specified address
+     * @param to address deposit address of token
+     * @param tokenId uint256 tokenID
+     */
+    function _mint(address to, uint256 tokenId) internal {
+        super._mint(to, tokenId);
+        emit Minted(to, tokenId);
     }
 
-     /**
-      * @dev Parses bytes and extracts a uint value
-      * @param data bytes From where to extract the index
-      * @return result the converted address
-      */
-    function bytesToUint(bytes memory data) internal pure returns (uint) {
-        require(data.length <= 256, "slicing out of range");
-        return abi.decode(data, (uint));
-     }
-
-     /**
-      * @dev Parses a uint and returns the hex string
-      * @param payload uint
-      * @return string the corresponding hex string
-      */
-     function uintToHexStr(uint payload) internal pure returns (string memory) {
-       if (payload == 0)
-         return "0";
-       // calculate string length
-       uint i = payload;
-       uint length;
-
-       while (i != 0) {
-         length++;
-         i = i >> 4;
-       }
-       // parse byte by byte and construct the string
-       i = payload;
-       uint mask = 15;
-       bytes memory result = new bytes(length);
-       uint k = length - 1;
-
-       while (i != 0) {
-         uint curr = (i & mask);
-         result[k--] = curr > 9 ? byte(55 + uint8(curr)) : byte(48 + uint8(curr));
-         i = i >> 4;
-       }
-
-       return string(result);
-     }
-
-    function equalBytes(bytes memory a, bytes memory b) internal pure returns (bool) {
-        if (a.length == b.length) {
-            for (uint i = 0; i < a.length; i++) {
-                if (a[i] != b[i]) {
-                    return false;
-                }
-            }
-            return true;
+    /**
+     * Computes the asset hash from the given params and verifies if its present in the AssetManager
+     * @param to address deposit address of the token
+     * @param properties bytes[] properties of the each proof
+     * @param values bytes[] value associated with each property
+     * @param salts bytes[] salt associated with each property
+     */
+    function _verifyAsset(address to, bytes[] memory properties, bytes[] memory values, bytes32[] memory salts) internal view returns (bool) {
+        require(to != address(0), "not a valid address");
```

Comment:
I don't think we need to catch this use case on chain. It will fail at a later point anyway.

Reply:
I've reopened this until it is addressed in the PR

Reply:
always fail-fast :)

Reply:
I have removed this check for now, but I would love it be there

Reply:
fail fast for extra gas is not worth it IMO. 99.999% of times this will be not be triggered but it will still cost people gas every time.